engine/src/actions/movement.rs

```diff
@@ -86,19 +115,94 @@ mod tests {
 
     #[test]
     fn go_up() {
-        expect_position(0.0, 0.0, up, 0.0, -1.0);
+        expect_position(0.0, 1.0, up, 0.0, 0.0);
         expect_position(1.0, 1.0, up, 1.0, 0.0);
     }
 
+    #[test]
+    fn move_to_occupied_position() {
+        let world: &mut World = &mut World::new();
+        let entity1 = world.register(Entity::new(0, Coordinate::new(1.0, 1.0)));
+        let entity2 = world.register(Entity::new(1, Coordinate::new(1.0, 2.0)));
+
+        let action = down(entity1.id);
+        let result = process(world, action);
+        assert!(result.is_err());
+        assert_eq!(result.err(), Some(EngineError::PositionOccupied(1.0, 2.0)));
+
+        let new_entity = world.get_entity(entity1.id).unwrap();
+        assert!(new_entity.coord.is_at_x(1.0));
+        assert!(new_entity.coord.is_at_y(1.0));
+
+        let new_entity2 = world.get_entity(entity2.id).unwrap();
+        assert!(new_entity2.coord.is_at_x(1.0));
+        assert!(new_entity2.coord.is_at_y(2.0));
+    }
+
+    #[test]
+    fn world_ends_error() {
+        const SIZE_X: i8 = 8;
+        const SIZE_Y: i8 = 8;
+        let world: &mut World = &mut World::create(SIZE_X as usize, SIZE_Y as usize);
+
+        expect_error(
+            world,
+            0.0,
+            0.0,
+            up,
+            EngineError::OutOfMapCoordinate(0.0, -1.0),
+        );
+        expect_error(
+            world,
+            0.0,
+            0.0,
+            left,
+            EngineError::OutOfMapCoordinate(-1.0, 0.0),
+        );
+        expect_error(
+            world,
+            f32::from(SIZE_X),
+            f32::from(SIZE_Y),
+            right,
+            EngineError::OutOfMapCoordinate((SIZE_X + 1).into(), SIZE_Y.into()),
+        );
+        expect_error(
+            world,
+            f32::from(SIZE_X),
+            f32::from(SIZE_Y),
+            down,
+            EngineError::OutOfMapCoordinate(SIZE_X.into(), (SIZE_Y + 1).into()),
+        );
+    }
+
     fn expect_position(x: f32, y: f32, f: fn(i32) -> ActionData, ex: f32, ey: f32) {
         let world: &mut World = &mut World::new();
         let entity = world.register(Entity::new(0, Coordinate::new(x, y)));
 
         let left = f(entity.id);
-        process(world, left);
+        assert!(process(world, left).is_ok());
 
         let new_entity = world.get_entity(entity.id).unwrap();
         assert!(new_entity.coord.is_at_x(ex));
         assert!(new_entity.coord.is_at_y(ey));
     }
+
+    fn expect_error(
+        world: &mut World,
+        x: f32,
+        y: f32,
+        f: fn(i32) -> ActionData,
+        error: EngineError,
+    ) {
+        let entity = world.register(Entity::new(0, Coordinate::new(x, y)));
+
+        let op = f(entity.id);
+        let result = process(world, op);
+        assert!(result.is_err());
```

Comment:
It seems like this assertion is redundant with the one that follows?

Reply:
👍 

Remains of TDDing that part. Well, maybe it's not the case, but because the compiler is so hard to please, I usually do baby steps.

Reply:
I know the feeling. I do these things myself all the time. Incrementalism means that this sort of thing will happen from time to time, but the benefits pay off 👍